packages/remirror__extension-file/src/file-extension.tsx

```diff
@@ -0,0 +1,355 @@
+        ...extra.defaults(),
+        url: { default: '' },
+        fileName: { default: '' },
+        fileType: { default: '' },
+        fileSize: { default: 0 },
+        error: { default: null },
+      },
+      ...override,
+      selectable: false,
+      draggable: true,
+      atom: true,
+      content: '',
+      parseDOM: [
+        {
+          tag: '[data-file]',
+          priority: ExtensionPriority.Low,
+          getAttrs: (dom) => {
+            const anchor = dom as HTMLAnchorElement;
+            const url = anchor.getAttribute('href');
+            const fileName = anchor.getAttribute('data-filename');
+            const fileType = anchor.getAttribute('data-filetype');
+            const fileSize = anchor.getAttribute('data-filesize');
+
+            return {
+              ...extra.parse(dom),
+              url,
+              fileName,
+              fileType,
+              fileSize,
+            };
+          },
+        },
+      ],
+      toDOM: (node) => {
+        const { url, ...rest } = omitExtraAttributes(node.attrs, extra);
+        const attrs = {
+          ...extra.dom(node),
+          ...rest,
+          href: url,
+          'data-file': '',
+          'data-filename': node.attrs.fileName,
+          'data-filetype': node.attrs.fileType,
+          'data-filesize': node.attrs.fileSize,
+        };
+
+        return ['a', attrs];
+      },
+    };
+  }
+
+  createExternalPlugins(): ProsemirrorPlugin[] {
+    return [placeholderPlugin];
+  }
+
+  createPasteRules(): PasteRule[] {
+    return [
+      {
+        type: 'file',
+        fileHandler: (props) => {
+          let pos: number | undefined;
+
+          if (props.type === 'drop') {
+            pos = props.pos;
+          }
+
+          for (const file of props.files) {
+            this.startLoadFile(file, pos);
+          }
+
+          return true;
+        },
+      },
+    ];
+  }
+
+  @command()
+  uploadFile(files: File[]): CommandFunction {
+    return () => {
+      for (const file of files) {
+        this.startLoadFile(file);
+      }
+
+      return true;
+    };
+  }
+
+  @command()
+  updateFile(pos: number, attrs: FileAttributes): CommandFunction {
+    return ({ tr, dispatch }) => {
+      dispatch?.(tr.setNodeMarkup(pos, undefined, attrs));
+      return true;
+    };
+  }
+
+  @command()
+  insertFile(attributes: FileAttributes, selection?: PrimitiveSelection): CommandFunction {
+    return ({ tr, dispatch }) => {
+      const { from, to } = getTextSelection(selection ?? tr.selection, tr.doc);
+      const node = this.type.create(attributes);
+
+      dispatch?.(tr.replaceRangeWith(from, to, node));
+
+      return true;
+    };
+  }
+
+  @command()
+  deleteFile(pos: number): CommandFunction {
+    return ({ tr, state, dispatch }) => {
+      const node = state.doc.nodeAt(pos);
+
+      if (node && node.type === this.type) {
+        if (dispatch) {
+          dispatch(tr.delete(pos, pos + 1).scrollIntoView());
```

Comment:
Think we should provide an `onFileDeleted` or maybe even `tr.setMeta`, so a user can delete files from remote storage when the node is removed from the document.

Reply:
```suggestion
          dispatch?.(tr.delete(pos, pos + 1).scrollIntoView());
```

Reply:
> or maybe even `tr.setMeta`

I don't understand this part. Could you give more details about this?

Reply:
I was thinking if we set meta on the transaction, then a user could write a plugin to detect those transactions and perform additional behaviour - i.e. deleting the actual file from their S3 bucket etc.

```ts
tr.setMeta(this.name, { action: 'DELETE', url: node.attrs.url })
dispatch?.(tr.delete(pos, pos + 1).scrollIntoView());
```

We could even expose an `onDeleteFile` handler option for this use case, [similar to how `onUpdateLink` works](https://github.com/remirror/remirror/blob/beta/packages/remirror__extension-link/src/link-extension.ts#L644-L662).

Reply:
I added a new `onDeleteFile` handler. Hope this option can cover most use cases. If users have some other requirement, then maybe they could override or wrap the `deleteFile` command.